### PR TITLE
add c++ gradient for op: Pow

### LINKED ISSUE #15239: No gradient defined for op: Pow
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 x64
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.4
- **Bazel version**: N/A
- **Python version**: None
- **CUDA/cuDNN version**: None
- **GPU model and memory**: None
- **Exact command to reproduce**: -

### Describe the problem
It seems there is no gradient defined for the Pow operation in the C++ API.

I am actually transferring this issue from https://github.com/migueldeicaza/TensorFlowSharp/issues/187. Similar to the case of Select (#14845), it seems there is also no gradient for the Pow operation in the C++ API.

### PR DESCRIPTION
**Pull request by facaiy**
Fix  #15239.

### How to test

+ [x] add test case.
+ [ ] pass all tests.

### CODE REVIEW

**Comment by caisq** [comment]
@tensorflow-jenkins test this please

**Comment by facaiy** [comment]
Thanks, @caisq . Git failed to fetch github branch. I think it's an unrelated failure.

File: tensorflow/cc/gradients/math_grad.cc [hunk 1]
```diff
@@ -473,6 +473,39 @@ Status AddNGrad(const Scope& scope, const Operation& op,
...
+  auto grad = grad_inputs[0];
+  // grad * y * pow(x, y - 1)
+  auto one = Cast(scope, Const(scope, 1.0), y.type());
+  auto gx_1 = Mul(scope,
+                  Mul(scope, grad, y),
+                  Pow(scope, x, Sub(scope, y, one)));
+  // Avoid false singularity at x = 0
+  DataType x_dtype = x.type();
+  auto zero = Cast(scope, Const(scope, 0.0), x_dtype);
+#define RETURN(comp)  \
```
Comment (suharshs) [inline tensorflow/cc/gradients/math_grad.cc]: I find this easier to read without the macro. I think its ok just to duplicate the code in each clause. thanks

File: tensorflow/cc/gradients/math_grad_test.cc [hunk 2]
```diff
@@ -843,6 +843,15 @@ TEST_F(NaryGradTest, SquaredDifference) {
   RunTest({x1, x2}, {x1_shape, x2_shape}, {y}, {x1_shape});
 }
 
+TEST_F(NaryGradTest, Pow) {
+  TensorShape x_shape({3});
```
Comment (suharshs) [inline tensorflow/cc/gradients/math_grad_test.cc]: x_shape and y_shape can both just be made 'shape'

**Review by suharshs — CHANGES_REQUESTED** [review]
just some minor comments. thanks for the PR!

**Comment by facaiy** [comment]
Thank you, @suharshs . I have revised the codes as requested.

**Review by suharshs — APPROVED** [review]

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by facaiy** [comment]
Time out failure:  //tensorflow/python/keras:data_utils_test

**Comment by caisq** [comment]
The test failure is unrelated. Merging PR.

**Comment by drpngx** [comment]
Looks like this test is flaky.

**Comment by gunan** [comment]
This is causing a lot of issues on the dashboard.
I propose a rollback, and then we can figure out if the test is flaky or the new op is.
